### ISSUE TITLE
'unknown variable accessed' processing template_file shell script which contains other variables

### ISSUE THREAD

**Issue by dhermans** [hashicorp/terraform#5683]
Hi,

I have userdata scripts that I am processing via template_file and referencing in an AWS launch configuration - all good.

I have a slightly more complicated userdata that does some AWS discovery for self configuration, eg:
ID=$(curl -s http://169.254.169.254/latest/meta-data/instance-id)
AZ=$(curl -s http://169.254.169.254/latest/meta-data/placement/availability-zone)

then it tries to do something useful with $ID or $AZ

problem is terraform thinks this is one if it's variables specified in the 'vars' definition and fails:
- template_file.test-userdata-a: failed to render : 29:9: unknown variable accessed: AZ

is there a way to tell terraform to ignore any variables that aren't it's own?

Thanks and Regards,
Great Tool!

**Comment by dhermans** [comment]
+1 to this it seems to be only some variables it complains about.
I didn't realize you could do '$$' so i fixed the first occurence (line 29 in the above example) eg:
and one other and it works.
if [ $${AZ} = 'ap-southeast-2a' ] ; then

So why don't ALL references to $variable inside the template require $$ substitution ?

Thanks!

**Comment by jen20** [comment]
Hi @dhermans! It's somewhat hard to follow what is happening here - could you post an example of such a template? Thanks!

**Comment by dhermans** [comment]
Hi,
sorry - here is a clearer example ( hopefully )..

say i have a template_file resource defined:
resource "template_file" "sample-userdata" {
   template = "${file("sample-userdata.template")}"
   vars {
      SetupBucket = "${var.SetupBucket}"
      ConfigDir   = "${var.stackname}"
   }
}

which is referenced in a launch config or instance:
user_data = "${template_file.sample-userdata.rendered}"


file 'sample-userdata.template' references the variables for substitution, eg:
aws s3 cp s3://${SetupBucket}/${ConfigDir}/

this works perfectly, at runtime this gets rendered with the variables substituted.

The same userdata template though, being a shell script, has code like:
AZ=$(curl -s http://169.254.169.254/latest/meta-data/placement/availability-zone)
if [ $AZ == "ap-southeast-2" ] ; then
   xxx

When this is executed via 'terraform apply' this error

    template_file.test-userdata-a: failed to render : 29:9: unknown variable accessed: AZ

no doubt because terraform thinks 'AZ' is one if it's variables specified in the 'vars' definition.

You can fix this with two '$':
if [ $$AZ == "ap-southeast-2" ] ; then

but it seems inconsistent as other variable references in the template DON'T require double $ 

Any thoughts as to what is happening?

**Comment by glasser** [comment]
Possibly the same as #5016?

**Comment by mitchellh** [comment]
It is and all this should be fixed up now!

**Comment by dizeee** [comment]
It's not the same at all and it's not fixed. The example above https://github.com/hashicorp/terraform/issues/5683#issuecomment-200095441 shows that it's impossible to use shell variables along with terraform variables in a shell script template, which makes terraform templating not usable for EC2 user data scripts, which is absolutely frustrating. The code itself has no errors. And in https://github.com/hashicorp/terraform/issues/5016 there's an error in the code, it's just reported incorrectly.
